allennlp/data/dataset_readers/dataset_reader.py

```diff
@@ -335,5 +289,122 @@ def deserialize_instance(self, string: str) -> Instance:
+            except Timeout:
+                # Couldn't acquire the lock within the timeout specified, so we'll
+                # just log a warning and ignore the cache.
+                logger.warning(
+                    "Failed to acquire lock on dataset cache file within {}s",
+                    self.CACHE_FILE_LOCK_TIMEOUT,
+                )
+                cache_file = None
+
+        if cache_file is not None:
+            # Try to acquire lock on cache file.
+            cache_file_lock = FileLock(cache_file + ".lock", timeout=self.CACHE_FILE_LOCK_TIMEOUT)
+            try:
+                cache_file_lock.acquire()
+            except Timeout:
+                # Couldn't acquire the lock within the timeout specified, so we'll
+                # just log a warning and ignore the cache.
+                logger.warning(
+                    "Failed to acquire lock on dataset cache file within {}s",
+                    self.CACHE_FILE_LOCK_TIMEOUT,
+                )
+                cache_file = None
+
+        if cache_file is not None and os.path.exists(cache_file):
+            # If the file already exists and we've acquired the lock then it must
+            # be up-to-date. Therefore we can release the lock and start reading from it.
+            cache_file_lock.release()
+            logger.info("Reading instances from cache %s", cache_file)
+            with open(cache_file) as data_file:
+                for instance in self._multi_worker_islice(
+                    data_file, transform=self.deserialize_instance
+                ):
+                    yield instance
+        elif cache_file is not None:
+            # We've acquired the lock but the cache file doesn't exist. So we'll try
+            # writing to it.
+            if util.is_distributed() or (get_worker_info() and get_worker_info().num_workers):
+                # But we can't write to the cache if there's more than one process loading
+                # instances since each worker only receives a partial share of the instances.
+                logger.warning(
+                    "Can't cache data instances when there are multiple processes loading data"
+                )
+                # Release the lock, then just read the instances as normal.
+                cache_file_lock.release()
+                for instance in self._multi_worker_islice(self._read(file_path)):
+                    yield instance
+            else:
+                # Otherwise we can safely write to it.
+                try:
+                    with open(cache_file, "w") as data_file:
+                        for instance in self._multi_worker_islice(self._read(file_path)):
+                            data_file.write(self.serialize_instance(instance) + "\n")
+                            yield instance
+                except:  # noqa: E722
+                    # If anything went wrong, the cache file will be corrupted, so we should
+                    # remove it.
```

Comment:
If the computer loses power while this is going on, we still end up with a half-written file. I think it's safer to write to a temporary file (`NamedTemporaryFile`?) in the same directory, and then rename it when the writing process finishes successfully.

I also don't know what happens if we yield a few instances, and then an exception gets thrown outside of here. Does that get propagated here? Does this `except` block run in that case?

I was curious, so I tried it. Looks like we don't see the exception:
```
In [1]: def f():
   ...:     try:
   ...:         yield 1
   ...:         yield 2
   ...:         yield 3
   ...:     except:
   ...:         print("We are excepting")
   ...:

In [2]: def g():
   ...:     g = f()
   ...:     print(next(g))
   ...:     print(next(g))
   ...:     raise ValueError(0)
   ...:     print(next(g))
   ...:

In [3]: g()
1
2
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-3-5fd69ddb5074> in <module>
----> 1 g()

<ipython-input-2-a306c1db4465> in g()
      3     print(next(g))
      4     print(next(g))
----> 5     raise ValueError(0)
      6     print(next(g))
      7

ValueError: 0
```

Reply:
Hmm interesting. Yea I guess writing to a temp file would be better then 👌